File: glue/generate.cpp
#include "generate.h"

static bool SetupModule(wasm::ModuleInterface* writer) {
	try {
		glue::State state{ writer };

		glue::SetupHostImports(state);

		glue::InitializeState(state);

		glue::SetupHostBody(state);
		glue::SetupContextFunctions(state);
		glue::SetupMapFunctions(state);
		glue::SetupMemoryFunctions(state);

		state.module.close();
	}
	catch (const wasm::Exception& e) {
		str::PrintWLn(L"Exception: ", e.what());
		return false;
	}
	return true;
}

bool glue::Generate(const std::string& wasmPath, const std::string& watPath) {
	{
		/* produce the module as WASM */
		writer::BinaryWriter _writer;
		if (!SetupModule(&_writer))
			return false;
		const std::vector<uint8_t>& data = _writer.output();

		/* write the file out */
		std::fstream _out{ wasmPath, std::ios::out | std::ios::binary };
		_out.write(reinterpret_cast<const char*>(data.data()), data.size());
		str::PrintLn(u8"Generated WASM at: [", wasmPath, u8']');
	}

	{
		/* produce the module as WAT */
		writer::TextWriter _writer;
		if (!SetupModule(&_writer))
			return false;
		const std::u8string& data = _writer.output();

		/* write the file out */
		std::fstream _out{ watPath, std::ios::out | std::ios::binary };
		_out.write(reinterpret_cast<const char*>(data.data()), data.size());
		str::PrintLn(u8"Generated WAT at: [", wasmPath, u8']');
	}

	return true;
}

File: glue/generate.py
#!/usr/bin/env python3

import os

# ensure the generated directory exists
dirGenerated = os.path.join(os.path.split(__file__)[0], '../server/generated')
if not os.path.isdir(dirGenerated):
	os.mkdir(dirGenerated)

# setup the paths
programPath = os.path.abspath(os.path.join(dirGenerated, './make-glue.exe'))
wasmPath = os.path.abspath(os.path.join(dirGenerated, './glue-module.wasm'))
watPath = os.path.abspath(os.path.join(dirGenerated, './glue-module.wat'))

# compile the glue generator and produce the generator
print('compiling...')
if os.system(f'cd {os.path.split(__file__)[0]} &&'
			 ' clang++ -std=c++20'
			 ' -O1'
			 f' -o {programPath}'
			 ' -I../repos'
			 ' make-glue.cpp'
			 ' generate.cpp'
			 ' context-functions.cpp'
			 ' host-functions.cpp'
			 ' initialize-state.cpp'
			 ' blocks-functions.cpp'
			 ' memory-functions.cpp'

			 ' ../repos/wasgen/objects/wasm-module.cpp'
			 ' ../repos/wasgen/sink/wasm-target.cpp'
			 ' ../repos/wasgen/sink/wasm-sink.cpp'
			 ' ../repos/wasgen/writer/text/text-base.cpp'
			 ' ../repos/wasgen/writer/text/text-module.cpp'
			 ' ../repos/wasgen/writer/text/text-sink.cpp'
			 ' ../repos/wasgen/writer/binary/binary-base.cpp'
			 ' ../repos/wasgen/writer/binary/binary-module.cpp'
			 ' ../repos/wasgen/writer/binary/binary-sink.cpp'
			) != 0:
	print('compilation failed')
	exit(1)
print(f'compiled at: [{programPath}]')

# execute the generator to produce its output
print('generating...')
if os.system(f'cd {dirGenerated} && make-glue.exe {wasmPath} {watPath}') != 0:
	print('generation failed')
	exit(1)
print('generation completed')
